genai/functions.go

```diff
@@ -0,0 +1,186 @@
+package genai
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+
+	pb "cloud.google.com/go/ai/generativelanguage/apiv1beta/generativelanguagepb"
+	"github.com/google/generative-ai-go/internal/support"
+)
+
+// A Tool is a piece of code that enables the system to interact with
+// external systems to perform an action, or set of actions, outside of
+// knowledge and scope of the model.
+type Tool struct {
+	// A list of `FunctionDeclarations` available to the model that can
+	// be used for function calling.
+	//
+	// The model or system does not execute the function. Instead the defined
+	// function may be returned as a [FunctionCall]
+	// with arguments to the client side for execution. The model may decide to
+	// call a subset of these functions by populating
+	// [FunctionCall][content.part.function_call] in the response. The next
+	// conversation turn may contain a
+	// [FunctionResponse][content.part.function_response]
+	// with the [content.role] "function" generation context for the next model
+	// turn.
+	FunctionDeclarations []*FunctionDeclaration
+}
+
+func (v *Tool) toProto() *pb.Tool {
+	if v == nil {
+		return nil
+	}
+	return &pb.Tool{
+		FunctionDeclarations: support.TransformSlice(v.FunctionDeclarations, (*FunctionDeclaration).toProto),
+	}
+}
+
+func (Tool) fromProto(p *pb.Tool) *Tool {
+	if p == nil {
+		return nil
+	}
+	return &Tool{
+		FunctionDeclarations: support.TransformSlice(p.FunctionDeclarations, (FunctionDeclaration{}).fromProto),
+	}
+}
+
+// FunctionDeclaration is structured representation of a function declaration as defined by the
+// [OpenAPI 3.03 specification](https://spec.openapis.org/oas/v3.0.3). Included
+// in this declaration are the function name and parameters.
+// Combine FunctionDeclarations into Tools for use in a [ChatSession].
+type FunctionDeclaration struct {
+	// Required. The name of the function.
+	// Must be a-z, A-Z, 0-9, or contain underscores and dashes, with a maximum
+	// length of 63.
+	Name string
+	// Required. A brief description of the function.
+	Description string
+	// Optional. Describes the parameters to this function.
+	Parameters *Schema
+	// If set the Go function to call automatically. Its signature must match
+	// the schema. Call [NewCallableFunctionDeclaration] to create a FunctionDeclaration
+	// with schema inferred from the function itself.
+	Function any
+}
+
+func (v *FunctionDeclaration) toProto() *pb.FunctionDeclaration {
+	if v == nil {
+		return nil
+	}
+	return &pb.FunctionDeclaration{
+		Name:        v.Name,
+		Description: v.Description,
+		Parameters:  v.Parameters.toProto(),
+	}
+}
+
+func (FunctionDeclaration) fromProto(p *pb.FunctionDeclaration) *FunctionDeclaration {
+	if p == nil {
+		return nil
+	}
+	return &FunctionDeclaration{
+		Name:        p.Name,
+		Description: p.Description,
+		Parameters:  (Schema{}).fromProto(p.Parameters),
+	}
+}
+
+// NewCallableFunctionDeclaration creates a [FunctionDeclaration] from a Go
```

Comment:
What's the significance of "Callable" here? Are there also non-callable function declarations?

Reply:
I mean "automatically callable." As in, the client will invoke the function for you, instead of returning a FunctionCall Part and having you provide the result. (The automatic calling isn't implemented yet.)

Reply:
> After all, the type is FunctionDeclaration, so NewFunctionDeclaration for the constructor sounds logical

True, but usually a function named `NewFoo` does about the same thing as `&Foo{...}`. Here the `New` function also makes the function automatically callable.